src/osx/cocoa/dataview.mm

```diff
@@ -2892,34 +2894,67 @@ void UpdateWithRow(int row)
         //else: can't change font if the cell doesn't have any
     }
 
-    if ( attr.HasColour() && [cell backgroundStyle] == NSBackgroundStyleLight )
+    // We don't apply the text or background colours if the cell is selected.
+    if ( [cell backgroundStyle] == NSBackgroundStyleLight )
     {
-        // we can set font for any cell but only NSTextFieldCell provides
-        // a method for setting text colour so check that this method is
-        // available before using it
-        if ( [cell respondsToSelector:@selector(setTextColor:)] &&
-                [cell respondsToSelector:@selector(textColor)] )
+        if ( attr.HasColour() )
         {
-            if ( !data->GetOriginalTextColour() )
+            // we can set font for any cell but only NSTextFieldCell provides
+            // a method for setting text colour so check that this method is
+            // available before using it
+            if ( [cell respondsToSelector:@selector(setTextColor:)] &&
+                    [cell respondsToSelector:@selector(textColor)] )
             {
-                // the cast to (untyped) id is safe because of the check above
-                data->SaveOriginalTextColour([(id)cell textColor]);
+                if ( !data->GetOriginalTextColour() )
+                {
+                    // the cast to (untyped) id is safe because of the check above
+                    data->SaveOriginalTextColour([(id)cell textColor]);
+                }
+
+                colText = attr.GetColour().OSXGetNSColor();
             }
+        }
 
-            colText = attr.GetColour().OSXGetNSColor();
+        if ( attr.HasBackgroundColour() )
+        {
+            // Use the same logic as the text colour check above
+            if ( [cell respondsToSelector:@selector(setBackgroundColor:)] &&
+                    [cell respondsToSelector:@selector(backgroundColor)] )
+            {
+                if ( !data->GetOriginalBackgroundColour() )
+                    data->SaveOriginalTextColour([(id)cell backgroundColor]);
+
+                colBack = attr.GetBackgroundColour().OSXGetNSColor();
+            }
         }
     }
 
+
     if ( !font )
         font = data->GetOriginalFont();
     if ( !colText )
         colText = data->GetOriginalTextColour();
+    if ( !colBack )
+        colBack = data->GetOriginalBackgroundColour();
 
     if ( font )
         [cell setFont:font];
 
     if ( colText )
         [(id)cell setTextColor:colText];
```

Comment:
I've finally tested this under Mac and this seems wrong: we can have valid `colText` because of fallback assignment of the original text colour above for a cell which doesn't respond to `setTextColor:`. And, sure enough, I get exceptions thrown from `-[NSObject(NSObject) doesNotRecognizeSelector:]` now... I'll fix with an explicit `respondsToSelector:` check but I don't understand how could this work for you.

Reply:
I don't know how I didn't see those exceptions. Were you testing the sample?

Reply:
Yes, but with the modifications from #1681. This should be the case without them too though, as the code already uses non-default background colour, but I didn't have time to check this.

Moreover, in this version of the sample setting the background colour doesn't work correctly (even after fixing the problem mentioned [here](https://github.com/wxWidgets/wxWidgets/pull/1673#discussion_r358000066)) as it's not reset for the odd rows, as it should be: instead, all cells in the first column have grey background. I'm not sure why does this happen as `[cell setDrawsBackground:false]` is getting called -- but apparently this isn't sufficient.

If you could please try to have a look at this, it would be very welcome, otherwise I'll try to debug this later. TIA!

Reply:
So I dug into the first column issue, and I think it has to do with the fact there is an attributed string present in the cell and that it is also an `NSButtonCell` not an `NSTextCell`. The fact that it is a button cell immediately seems to break the font coloring as well, so the text color isn't yellow it is black.

I had a look through the rendering pipeline for the cell attributes on OSX, and I think that it would make sense to pass in the cell's `wxDataViewItemAttr` to the `MacRender` function inside its call on dataview.mm:1872 and use the MacRender function to set the cell attributes instead of the SetAttr function. Many of these attributes seem to be cell-type specific. This way the attributes are available at the creation of the attributed string and the cell content so the attributed string can be given them. This will allow things such as the text color to actually work on these non-NSTextCell cells, and possibly allow for things like strikethrough to be implemented (since that looks like it would require an attributed string in the cell).

Reply:
Ok, I think I narrowed down the odd behavior that is seen in the first column. What seems to be happening is that the NSButtonCell used doesn't have any color defined in the `backgroundColor` field, so it is NULL. When it reads that to save the color, it doesn't actually save any colors and so it will keep searching each future cell until it finds one that has a color defined. After the first pass through all the cells, it will then readback one of the background colors that has already been set. Therefore we get the background color we set on every cell.

I am not sure the best way of fixing this though.

Reply:
I could be missing something, but it seems that we just should get rid of `m_origBackgroundColour` and everything related to it entirely. I.e. just leave it null and call `[NSCell setDrawsBackground:false]` in this case. Wouldn't this fix the problem?

I.e. I have the impression that the code was just copied from the existing one dealing with the font, but the difference is that the cell does have some font originally whereas it doesn't have background colour nor does it need it because we can tell it to simply not use it.